src/tallies/filter_energy.cpp

```diff
@@ -0,0 +1,174 @@
+#include "openmc/tallies/filter_energy.h"
+
+#include "openmc/capi.h"
+#include "openmc/constants.h"  // For F90_NONE
+#include "openmc/mgxs_interface.h"
+#include "openmc/search.h"
+#include "openmc/settings.h"
+#include "openmc/xml_interface.h"
+
+namespace openmc {
+
+//==============================================================================
+// EnergyFilter implementation
+//==============================================================================
+
+void
+EnergyFilter::from_xml(pugi::xml_node node)
+{
+  auto bins = get_node_array<double>(node, "bins");
+  this->set_bins(bins);
+}
+
+void
+EnergyFilter::set_bins(gsl::span<const double> bins)
+{
+  // Clear existing bins
+  bins_.clear();
+  bins_.reserve(bins.size());
+
+  // Copy bins, ensuring they are valid
+  for (gsl::index i = 0; i < bins.size(); ++i) {
+    if (i > 0 && bins[i] <= bins[i-1]) {
+      throw std::runtime_error{"Energy bins must be monotonically increasing."};
+    }
+    bins_.push_back(bins[i]);
+  }
+
+  n_bins_ = bins_.size() - 1;
+
+  // In MG mode, check if the filter bins match the transport bins.
+  // We can save tallying time if we know that the tally bins match the energy
+  // group structure.  In that case, the matching bin index is simply the group
+  // (after flipping for the different ordering of the library and tallying
+  // systems).
+  if (!settings::run_CE) {
+    if (n_bins_ == data::num_energy_groups) {
+      matches_transport_groups_ = true;
+      for (gsl::index i = 0; i < n_bins_ + 1; ++i) {
+        if (data::rev_energy_bins[i] != bins_[i]) {
+          matches_transport_groups_ = false;
+          break;
+        }
+      }
+    }
+  }
+}
+
+void
+EnergyFilter::get_all_bins(const Particle* p, int estimator, FilterMatch& match)
+const
+{
+  if (p->g_ != F90_NONE && matches_transport_groups_) {
+    if (estimator == ESTIMATOR_TRACKLENGTH) {
+      match.bins_.push_back(data::num_energy_groups - p->g_);
+    } else {
+      match.bins_.push_back(data::num_energy_groups - p->g_last_);
+    }
+    match.weights_.push_back(1.0);
+
+  } else {
+    // Get the pre-collision energy of the particle.
+    auto E = p->E_last_;
+
+    // Bin the energy.
+    if (E >= bins_.front() && E <= bins_.back()) {
+      auto bin = lower_bound_index(bins_.begin(), bins_.end(), E);
+      match.bins_.push_back(bin);
+      match.weights_.push_back(1.0);
+    }
+  }
+}
+
+void
+EnergyFilter::to_statepoint(hid_t filter_group) const
+{
+  Filter::to_statepoint(filter_group);
+  write_dataset(filter_group, "bins", bins_);
+}
+
+std::string
+EnergyFilter::text_label(int bin) const
+{
```

Comment:
We should probably have error handling to ensure bin is not greater bin_.size() - 2